File: mobile/src/api/inventory.ts
import { api } from './client';

export type MovementType = 'purchase' | 'adjustment' | 'sale';
export type AdjustReason = 'damage' | 'loss' | 'correction' | 'expired';

export const inventoryApi = {
  movements: (limit = 20) =>
    api.get(`/inventory/movements?limit=${limit}`),

  addMovement: (body: {
    product_id: string;
    movement_type: MovementType;
    quantity: number;
    unit_cost_pesawas?: number;
    adjustment_sign?: '+' | '-';
    reason?: AdjustReason;
    notes?: string;
  }) => api.post('/stock/movements', body),

  audit: () => api.get('/inventory/audit'),

  confirmAudit: (body: {
    audit_id: string;
    signed_by: string;
    items: { product_id: string; actual_qty: number }[];
  }) => api.post('/inventory/audit/confirm', body),
};

File: backend/app/models/inventory.py
import uuid
from datetime import datetime, timezone

from sqlalchemy import DateTime, ForeignKey, Integer, String
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class StockMovement(Base):
    __tablename__ = "stock_movements"

    id: Mapped[str] = mapped_column(UUID(as_uuid=False), primary_key=True, default=lambda: str(uuid.uuid4()))
    shop_id: Mapped[str] = mapped_column(UUID(as_uuid=False), ForeignKey("shops.id", ondelete="CASCADE"), nullable=False, index=True)
    product_id: Mapped[str] = mapped_column(UUID(as_uuid=False), ForeignKey("products.id", ondelete="CASCADE"), nullable=False, index=True)
    movement_type: Mapped[str] = mapped_column(String(20), nullable=False)  # purchase / adjustment / sale
    quantity: Mapped[int] = mapped_column(Integer, nullable=False)
    unit_cost_pesawas: Mapped[int | None] = mapped_column(Integer, nullable=True)
    adjustment_sign: Mapped[str | None] = mapped_column(String(1), nullable=True)  # '+' or '-'
    reason: Mapped[str | None] = mapped_column(String(50), nullable=True)  # damage / loss / correction / expired
    notes: Mapped[str | None] = mapped_column(String(500), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), index=True)

    product: Mapped["Product"] = relationship("Product", back_populates="stock_movements")
